cpp/src/reader/aligned_chunk_reader.cc

```diff
@@ -0,0 +1,623 @@
+    }
+    if (compressor_ != nullptr) {
+        compressor_->~Compressor();
+        CompressorFactory::free(compressor_);
+        compressor_ = nullptr;
+    }
+    char *buf = time_in_stream_.get_wrapped_buf();
+    if (buf != nullptr) {
+        mem_free(buf);
+        time_in_stream_.clear_wrapped_buf();
+    }
+    cur_time_page_header_.reset();
+    buf = value_in_stream_.get_wrapped_buf();
+    if (buf != nullptr) {
+        mem_free(buf);
+        value_in_stream_.clear_wrapped_buf();
+    }
+    cur_value_page_header_.reset();
+}
+
+int AlignedChunkReader::load_by_aligned_meta(ChunkMeta *time_chunk_meta,
+                                             ChunkMeta *value_chunk_meta) {
+    int ret = E_OK;
+    time_chunk_meta_ = time_chunk_meta;
+    value_chunk_meta_ = value_chunk_meta;
+#if DEBUG_SE
+    std::cout << "AlignedChunkReader::load_by_meta, meta=" << *time_chunk_meta
+              << ", " << *value_chunk_meta << std::endl;
+#endif
+    /* ================ deserialize time_chunk_header ================*/
+    // at least, we can reader the chunk header and the first page header.
+    // TODO configurable
+    file_data_time_buf_size_ = 1024;
+    file_data_value_buf_size_ = 1024;
+    int32_t ret_read_len = 0;
+    char *time_file_data_buf =
+        (char *)mem_alloc(file_data_time_buf_size_, MOD_CHUNK_READER);
+    if (IS_NULL(time_file_data_buf)) {
+        return E_OOM;
+    }
+    ret = read_file_->read(time_chunk_meta_->offset_of_chunk_header_,
+                           time_file_data_buf, file_data_time_buf_size_,
+                           ret_read_len);
+    if (IS_SUCC(ret) && ret_read_len < ChunkHeader::MIN_SERIALIZED_SIZE) {
+        ret = E_TSFILE_CORRUPTED;
+        LOGE("file corrupted, ret=" << ret << ", offset="
+                                    << time_chunk_meta_->offset_of_chunk_header_
+                                    << "read_len=" << ret_read_len);
+    }
+    if (IS_SUCC(ret)) {
+        time_in_stream_.wrap_from(time_file_data_buf, ret_read_len);
+        if (RET_FAIL(time_chunk_header_.deserialize_from(time_in_stream_))) {
+        } else {
+            time_chunk_visit_offset_ = time_in_stream_.read_pos();
+        }
+    }
+    /* ================ deserialize value_chunk_header ================*/
+    ret_read_len = 0;
+    char *value_file_data_buf =
+        (char *)mem_alloc(file_data_value_buf_size_, MOD_CHUNK_READER);
+    if (IS_NULL(value_file_data_buf)) {
+        return E_OOM;
+    }
+    ret = read_file_->read(value_chunk_meta_->offset_of_chunk_header_,
+                           value_file_data_buf, file_data_value_buf_size_,
+                           ret_read_len);
+    if (IS_SUCC(ret) && ret_read_len < ChunkHeader::MIN_SERIALIZED_SIZE) {
+        ret = E_TSFILE_CORRUPTED;
+        LOGE("file corrupted, ret="
+             << ret << ", offset=" << value_chunk_meta_->offset_of_chunk_header_
+             << "read_len=" << ret_read_len);
+    }
```

Comment:
value_file_data_buf and time_file_data_buf should be freed (if not null) before each return statement of failure.

Reply:
Free memory on file read failure in `chunk_reader` and `aligned_chunk_reader`